syntax = "proto2";
package oneflow;

import "oneflow/core/common/data_type.proto";
import "oneflow/core/common/shape.proto";

enum NativeFlexType {
  kNativeFlexTypeBool = 1;
  kNativeFlexTypeInt64 = 2;
  kNativeFlexTypeUint64 = 3;
  kNativeFlexTypeDouble = 4;
  kNativeFlexTypeDataType = 5;
  kNativeFlexTypeString = 6;
  kNativeFlexTypeShape = 7;
}

message NativeFlexValueProto {
  oneof value {
    bool bool_val = 1; // field_numbe == kNativeFlexTypeBool
    int64 int64_val = 2; // field_numbe == kNativeFlexTypeInt64
    uint64 uint64_val = 3; // field_numbe == kNativeFlexTypeUint64
    double double_val = 4; // field_numbe == kNativeFlexTypeDouble
    DataType data_type_val = 5; // field_numbe == kNativeFlexTypeDataType
    string string_val = 6; // field_numbe == kNativeFlexTypeString
    ShapeProto shape_val = 7; // field_numbe == kNativeFlexTypeShape
  }
}

message ListFlexValueProto {
  repeated FlexValueProto flex_value = 1;
}

message StructFlexValueProto {
  map<string, FlexValueProto> field_name2flex_value = 1;
}

message FlexValueProto {
  oneof type {
    NativeFlexValueProto native_flex_value = 1;
    ListFlexValueProto list_flex_value = 2;
    StructFlexValueProto struct_flex_value = 3;
  }
}

enum FlexLabel {
  kFlexLabelRequired = 1;
  kFlexLabelOptional = 2;
}

message NativeFlexDefProto {
  required NativeFlexType native_flex_type = 1;
}

message FlexFieldDefProto {
  required FlexLabel label = 1;
  required FlexDefProto flex_def = 2;
  required string field_name = 3;
  optional FlexValueProto default_val = 5;
  optional string description = 6;
}

message StructFlexDefProto {
  map<string, FlexFieldDefProto> field_name2field = 1;
}

message ListFlexDefProto {
  optional FlexDefProto elem_flex_def = 1;
}

message FlexDefProto {
  oneof type {
    NativeFlexDefProto native_flex_def = 1;
    StructFlexDefProto struct_flex_def = 2;
    ListFlexDefProto list_flex_def = 3;
  }
}
